execution/evm/vm.go

```diff
@@ -68,6 +68,7 @@ type VM struct {
 	returnData       []byte
 	debugOpcodes     bool
 	dumpTokens       bool
+	inStaticCall     bool
```

Comment:
I think it might be better if we ensure no state writes at the state and event sink level. Rather than adding another contextual flag here:

- It will catch any state mutation we may not have thought about
- I think it's cleaner than managing this state in the EVM (with the various checks and resetting that is necessary)
- 'Context is bad' - we already have some but still.

However let's pin this thought (I'll PR some code that illustrates the alternative I have in mind) but there is useful stuff to be done in this PR anyway.